### ISSUE TITLE
[Improve] Add a listener mechanism for Paimon's commit

### ISSUE THREAD

**Issue by liming30** [apache/incubator-paimon#959]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

At present, Paimon's commit behavior is not perceived by the outside world, but in some scenarios, especially platform-based scenarios, we may need to monitor Paimon's commit behavior to decide when to perform a compaction.

You may ask why not use Paimon's automatic compaction? When we used Flink for streaming computing, `RocksDBStateBackend` would also perform compaction periodically, which would cause periodic spikes in CPU usage, which is very unfriendly to streaming computing. The behavior of compaction will compete with computing tasks for CPU, and usually the CPU usage of compaction is high, but the duration is short, resulting in low resource utilization. 

Therefore, separating the compaction behavior from the streaming job and letting an external platform take charge of the compaction will improve resource utilization and be more friendly to streaming computing.

### Solution

Add a listener for Paimon's commit behavior. This listener supports user-defined extensions, such as reporting to external services through the rest api.

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

**Comment by SteNicholas** [comment]
@liming30, does above mean to introduce the `CommitPolicy` interface to define the corresponding behavior of commit？

**Comment by liming30** [comment]
@SteNicholas I prefer an interface similar to `listener`. This mechanism should not modify the existing commit behavior, it is more like adding a `callback`.

**Comment by JingsongLi** [comment]
How about just a streaming reader? We can modify snapshot generation.

**Comment by liming30** [comment]
> How about just a streaming reader? We can modify snapshot generation.

@JingsongLi Sorry, I didn't quite understand what this meant. I think the more meaningful scenario is to remove the compaction behavior from the streaming job or the external service can decide when to initiate a `full-compaction` based on the commit information.

**Comment by JingsongLi** [comment]
> > How about just a streaming reader? We can modify snapshot generation.
> 
> @JingsongLi Sorry, I didn't quite understand what this meant. I think the more meaningful scenario is to remove the compaction behavior from the streaming job or the external service can decide when to initiate a `full-compaction` based on the commit information.

We already have https://paimon.apache.org/docs/master/maintenance/write-performance/#dedicated-compaction-job for streaming mode, it continuously modify snapshot generation and do compaction. Same way?

**Comment by liming30** [comment]
> We already have https://paimon.apache.org/docs/master/maintenance/write-performance/#dedicated-compaction-job for streaming mode, it continuously modify snapshot generation and do compaction. Same way?

@JingsongLi Yes, running `dedicated-compaction-job` in stream mode can decouple compaction from streaming-job and continuously do compaction. But I may not be willing to use this solution, there are two considerations:

1. From the perspective of `resource utilization`, `dedicated-compaction-job` is an intermittent job (depending on the generation of snapshot, which depends on the production of checkpoint), which cannot effectively utilize resources;

2. From the perspective of `compaction trigger timing`, I roughly read the code related to `CompactorSink`. It seems that compaction will be triggered when a new snapshot is generated, and some other factors are not considered, such as write amplification, space amplification, etc. Frequent compactions may also have a negative impact on storage.

Finally, I expect that there can be an external service and a resident `compaction-job`. The `compaction-job` can handle the compaction task of different tables.  This compaction-job can handle the compaction of different tables, and external services distribute compaction tasks to the `compaction-job` according to different strategies (such as table space amplification, write amplification, etc.), which can improve resource utilization and configure more flexible compaction strategies.

**Comment by FangYongs** [comment]
@liming30 The external service you mentioned sounds like a `table service` in hudi? From a requirements perspective, what you need is not an listener, but a 'Table Service' that can collect enough information to manage multiple tables in Paimon?

**Comment by liming30** [comment]
> @liming30 The external service you mentioned sounds like a `table service` in hudi? From a requirements perspective, what you need is not an listener, but a 'Table Service' that can collect enough information to manage multiple tables in Paimon?

@FangYongs Yes, `external service` is similar to `table service` and is mainly used to manage tables in a larger scope, such as clusters, tenants, etc. We need to provide enough information for it.

**Comment by JingsongLi** [comment]
What I can understand is that listener can reduce the polling of the file system.

**Comment by JingsongLi** [comment]
A table service looks good, how to execute compaction? Batch jobs?
